### ISSUE TITLE
Three Questions About Memtorch

### ISSUE THREAD

**Issue by Jize-W** [coreylammie/MemTorch#20]
Hello, I have 3 problems here. 

1. When I apply the 'device-to-device variability', I find that when the sigma reaches very high, r_off may be approximately equal to r_on, thus there will be an error 'ZeroDivisionError: float division by zero'. 
2. Does Memtorch have the function of measuring power consumption?
3. When I apply the 'device failure', even if I change '.view(-1)' to '.reshape(-1)' in DeviceFaults.py, the error 'RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.' still exists.

How to tackle with these problems?

**Comment by Jize-W** [comment]
@coreylammie

**Comment by coreylammie** [comment]
Hi Jize-W,

Thank you posting your questions/problems! I have created a new issue for each problem that you have listed, that I intend on addressing in the near future. I am currently developing/implementing modular crossbar tile and quantization support, that I intend on including in the 1.1.0 release, so I'll try to address (1) and (3), and bundle their fixes in the same release.

1.  Thank you for informing me of this behavior. I have created a standalone issue outlining this problem, and I intend on adding a check to ensure that the 'ZeroDivisionError: float division by zero' is not raised. As a side note- in literature, r_off is commonly sampled from a normal distribution with double the standard deviation of that used to sample r_on when device-device variability is modelled simplistically. As you have noted, this may cause problems when the standard deviations used are quite large. Although all examples sample r_off from a normal distribution with 2\sigma and r_on from a normal distribution with /sigma, you are free to change this ratio how you see it. 
2. Not currently, although this is definitely on my TODO list. I currently intend on adding modular tile and quantization support to the 1.1.0 release, and adding power/area/timing estimates to a later release, such as 1.2.0.
3. I am aware of this issue, and it will be fixed when I merge the _tile_ branch to master. 

Kind Regards,

Corey.

**Comment by Jize-W** [comment]
**Thanks a lot for your detailed answers and the new implementation! 

However, when I try the new version, a new error occurs in the procedure of `model.tune_()`. The specific location is "mn/Conv2d.py" in _forward_, line122-`out[batch] = out_.view(size=(1, self.out_channels, output_dim[0], output_dim[1]))`".**

**I didn't make any change to my code, which can run through the previous version. 
It says "RuntimeError: shape '[1, 4, 32, 32]' is invalid for input of size 3136".**  

(Here are some of my explorations, which may not help...)

I find that the error changes with the default value of input_shape in `tune(self, input_batch_size=8, input_shape=32): `
input_shape=32        [1, 4, 32, 32]         3136=1*4*28*28
input_shape=28        [1, 4, 28, 28]         2304=1*4*24*24
input_shape=14        [1, 4, 14, 14]         400  =1*4*10*10

input_shape=a          [1, 4, a, a]                        =1*4*b*b
where b=a-4

If I just arbitrarily changed `out[batch] = out_.view(size=(1, self.out_channels, output_dim[0], output_dim[1]))` to `out[batch] = out_.view(size=(1, self.out_channels, output_dim[0]-4, output_dim[1]-4))`, it says "RuntimeError: The expanded size of the tensor (32) must match the existing size (28) at non-singleton dimension 2.  Target sizes: [4, 32, 32].  Tensor sizes: [4, 28, 28]"

I don't know what's wrong with the new version, maybe the transformation of dimensions?

**Comment by coreylammie** [comment]
No problem at all! I've just pushed a hot-fix to master that should fix dimension mismatch errors for convolutional layers with non-zero padding and modified unit tests to check for instances were non-zero padding is used. If this does not resolve your issue, could you please specify the parameters of the original torch.nn.Conv2d layer (in_channels, out_channels, kernel_size, stride, and padding), and I can further investigate?

**Comment by Jize-W** [comment]
It seems that the error of dimension mismatch has been solved, but a new error (conversion between CPU and CUDA) has been reported. Here is the code that I am using. Maybe you can try this code to find out the problems?
[mnist_test.zip](https://github.com/coreylammie/MemTorch/files/5899388/mnist_test.zip)

**Comment by coreylammie** [comment]
Thanks for sharing your code! It appears that _reg.coef__ and _reg.intercept__ can return single element arrays with a varying number of dimensions, where _reg = linear_model.LinearRegression(fit_intercept=True).fit(output, legacy_output)_. In your particular instance, _reg.coef__ returned a 2D single element array, so _reg.coef_[0]_ pointed to a 1D single element array, and not a numeric value, which caused the error.

I've pushed another hot-fix to master that extracts the element of a single-element N-dimensional arrays for _reg.coef__ and _reg.intercept__, so it should now be robust to all cases. I can confirm that your code should now be functional using the latest version- I added a pandas import (import pandas as pd), and all cells in your Jupyter notebook successfully executed on my machine.

**Comment by Jize-W** [comment]
Thanks! There's something else:

1. About the 'r_on and r_off are indistinguishable' problem, you raised an exception 'r_off and r_on values are indistinguishable.' in StochasticParameter.py.  

In my opinion, though there will be a distribution overlap between r_on and r_off when \sigma reaches very large, for a single memristor, its r_off must be much larger than r_on, which is decided by the physical characteristics of the memristor. That is to say, from the perspective of modeling, the 'r_off and r_on values are indistinguishable' case shouldn't exist. 

But when we generate r_on values and r_off values from two normal distributions, we don't consider the fact that r_off is much larger than r_on, which leads to the 'indistinguishable' error.

So is it possible/ more reasonable to limit the difference between one r_on value and its corresponding r_off value ( They are of the same memristor) to be greater than a certain threshold, when they are generated from the specified distribution, instead of raising an error?

2. Maybe you can set an option to let the user choose whether to output the network parameters in the patching/tuning process. Because when the network is relatively large, there are too many output messages, which is not very convenient.

3. In the file ./bh/crossbar/Tile.py

In line 41 & 43:  `torch.tensor(new_array)` => `new_array.clone().detach()`

In line 114:  `tiles = torch.tensor([np.array(tile.array) for tile in tiles])` => `tiles = torch.tensor([np.array(tile.array.cpu()) for tile in tiles])`

**Comment by coreylammie** [comment]
No problem at all. Thank you for all of your suggestions! These issues/enhancements should be fairly straight-forward to implement, so I'll create a new branch and implement them at some stage tomorrow.

1. I agree. I have added a new issue (#27) detailing this. In cases were `math.isclose(r_off, r_on)` is true, _r_off_ and _r_on_ should be resampled, where _rel_tol_ or _abs_tol_ can be specified. This should avoid any nuances when manually shifting sampled values. A side note- although stochastic parameters are usually sampled from normal distributions, all PyTorch distributions from _torch.distributions.distribution.Distribution_ are supported by the _StochasticParameter_, so other distributions can be used.
2. I have added a new issue (#25) detailing this. At some stage in the future I plan on adding customizable degrees of 'verbose output'.
3. I have added a new issue (#26) detailing your suggested changes to make.
